### ISSUE TITLE
False positive unhashable set entry for frozen pydantic models

### ISSUE THREAD

**Issue by Vegemash** [microsoft/pyright#4168]
**Describe the bug**
Pyright reports that frozen pydantic types are unhashable and thus cannot be contained in a set or be dict keys. They have the `__hash__` function and can be used in these areas.

**To Reproduce**
See code below

**Expected behavior**
If a class implements `__hash__` it should work in sets and as dict keys and not throw an error from pyright.

**Screenshots or Code**
pydantic version 1.10.2
```python
from pydantic import BaseModel


class T(BaseModel):
    a: str = "default"

    class Config:
        frozen = Truel


assert hasattr(T, "__hash__")
example_set: set[T]
example_set = {T(), T()}
assert len(example_set) == 1
```

**VS Code extension or command-line**
Command-line

**Comment by erictraut** [comment]
Pydantic uses a non-standard way of indicating that a class is frozen. Standards-based type checkers like pyright therefore don't have any knowledge that a pydantic model implements a `__hash__` method.

Pydantic makes use of `dataclass_transform` (defined in [PEP 681](https://peps.python.org/pep-0681/)). The `dataclass_transform` mechanism supports `frozen` classes if the dataclass mechanism uses a class decorator approach, but pydantic uses a metaclass approach to define `BaseModel`, so there's no way to specify `frozen` in a standard way.

[This recent pyright enhancement](https://github.com/microsoft/pyright/issues/4145) introduced checks for hashability. Prior to this, the hashability of pydantic classes was not an issue.

I'll need to think more about how to handle this. Perhaps we need to simply assume that all classes that use `dataclass_transform` with a metaclass or base class approach all synthesize a `__hash__` function.

@samuelcolvin, @debonte, let me know if you have any thoughts here.

**Comment by erictraut** [comment]
@debonte, I've prepared [a PR](https://github.com/microsoft/pyright/pull/4169) that implements the change I proposed above. If we can come up with a better solution to this issue, I'll abandon the PR. If we can't come up with a better solution, then we have this one to fall back on.

**Comment by debonte** [comment]
> The dataclass_transform mechanism supports frozen classes if the dataclass mechanism uses a class decorator approach, but pydantic uses a metaclass approach to define BaseModel, so there's no way to specify frozen in a standard way.

I think that's incorrect. Or maybe I'm misunderstanding what you're saying. The PEP actually shows an example of how to use frozen with metaclasses: https://peps.python.org/pep-0681/#id3.

And this code works:

```python
from typing_extensions import dataclass_transform


@dataclass_transform()
class ModelMeta(type):
    def __new__(
        cls,
        name,
        bases,
        namespace,
        *,
        frozen: bool = False,
        eq: bool = True,
    ):
        ...
        
class ModelBase(metaclass=ModelMeta): ...
  
class CustomerModel(
    ModelBase,
    frozen=True,
):
    id: int
    name: str
    

example_set = {CustomerModel(1, "foo"), CustomerModel(2, "bar")}  # No error as long as frozen and eq are both True
```

**Comment by erictraut** [comment]
Ah, nice. It looks like pydantic already supports this form. I'll close my PR then.

@Vegemash, you can make this work by changing your code as follows:

```python
class T(BaseModel, frozen=True):
    a: str = "default"
```

PEP 681 doesn't support pydantic's `Config` nested class, so this isn't supported by pyright.

**Comment by Vegemash** [comment]
@erictraut That works! Thank you
